### PR TITLE
Feature/xinnanli/aff bug

### LINKED ISSUE #578: Primary Affiliations to Accounts w/o Recordtypes cause Null error
If I fill in Primary Business Organization (or any other Contact "primary" field, HEDA will automatically create a matching affiliation record with the Primary checkbox checked.

If I try to change the Primary Business Organization to another Account or clear the field entirely, I get an error.  If I try to uncheck the Primary checkbox on the matching Affiliation record, I get the same error. 

Error is "execution of AfterUpdate caused by: System.NullPointerException: Argument cannot be null. (hed) : [] (hed)"

I suspect HEDA is trying to match the Account record type (blank) to HEDA settings for Affiliation Mappings and coming up with Null. Currently the only way to get around the error is to manually delete the Affiliation record and then change/clear the value in Primary Business Organization.

While we encourage Record Types on all accounts, it's possible that existing account data (pre-HEDA) may not have a record type and could cause these errors. I also suspect any mismatches between Account Record Type setup and HEDA Affiliation mappings could also cause problems.

### PR DESCRIPTION
**Pull request by xinnanlisf**
# Critical Changes

# Changes
- We've added a new setting in HEDA Settings, Enable Record Type Validation, to prevent record type errors when users specify an Account as a Contact's Primary Affiliation. When the setting is selected, we now validate that affiliated Accounts have a record type and that all Account Record Type values are correctly mapped to their respective Contact Primary Affl Fields in Affiliation Mappings. If validation fails, an error prevents the Affiliation from being saved. (See <a href="https://powerofus.force.com/articles/Resource/HEDA-Create-Accounts-and-Affiliations">Create Primary Affiliation Mappings</a> to see the correlation between Account record types and Contact Primary Affl fields.) 
- In an existing org, this validation is off by default. To turn it on, you must select Enable Record Type Validation in HEDA Settings, under Affiliations | Settings.

# Issues Closed
Fixes #578
# New Metadata

# Deleted Metadata

# Testing Notes
1. Create accounts without record types before deploying HEDA
2. Deploy HEDA
3. Create a contact and use one of the testing account as primary field (Business org etc).
4. Go to HEDA settings and check Affiliation Record Type enforced setting
4. Change the primary field to another account without record type/ uncheck primary checkbox on the new affiliation created.
5. Make sure the proper error message is shown.
6. Create another contact and use another account as primary field (Business or etc). 
7. Go to setting page and modify the affiliation account mappings to not matching existing record types.
8. Change the primary field to another account without record type/ uncheck primary checkbox on the new affiliation created.
9. Make sure the proper error message is shown.

### CODE REVIEW

**Comment by annejackson** [comment]
Issue found (with some others that @xinnanlisf  found):
In a contact
I had a primary affiliation to a business organization record type account
In related list->I Edited that affiliation and choose an account without a record type (and left primary checked) and pressed Save button- Got an error.

TDTM_Affiliation: execution of AfterUpdate caused by: System.NullPointerException: Argument cannot be null. Class.AFFL_AccChange_TDTM.run: line 96, column 1 Class.TDTM_TriggerHandler.runClass: line 145, column 1 Class.TDTM_TriggerHandler.run: line 75, column 1 Class.TDTM_Global_API.run: line 61, column 1 Trigger.TDTM_Affiliation: line 33, column 1

edit affilation error

We also talked about this concern:
If we are putting a new enforcement in place for recordtypes on those primary fields, it's possible that existing integrations will break if a customer has things set up incorrectly. I'm an advocate for having a HEDA setting for "Enforce record type mappings when populating mapped fields". Even if we check it by default during the upgrade, we have a path do disable it for customers to get back up and running until they fix their integration. I see it as similar to the "enforce picklist values" feature in platform...good to have but can't force it on orgs. 

Some ideas:
enforce record type mappings when populating mapped fields- for trial orgs -default checked to enforce
for brand new install/orgs- default checked to enforce,
Existing heda orgs- defaulted to not checked to enforce 
**all have the options to modify this setting in heda settings

**Comment by xinnanlisf** [comment]
Hey @lparrott, this most part of this PR should be ready for review now. I might need to work on 1 or 2 unit tests thou. Thank you! @annejackson  feel free to start testing if you want. :)

File: src/package.xml [hunk 1]
```diff
@@ -552,6 +556,8 @@
         <members>stgTitleNewReciSetting</members>
         <members>stgTitleReciMethod</members>
         <members>stgToastError</members>
+        <members>afflAccoutMappingError</members>
```
Comment (lparrott) [inline src/package.xml]: Looks like these two labels got added twice.

File: src/classes/UTIL_CustomSettingsFacade.cls [hunk 2]
```diff
@@ -138,6 +138,7 @@ public without sharing class UTIL_CustomSettingsFacade {
         settings.Disable_Preferred_Email_Enforcement__c = mySettings.Disable_Preferred_Email_Enforcement__c;
         settings.Household_Account_Naming_Format__c = mySettings.Household_Account_Naming_Format__c;
         settings.Automatic_Household_Naming__c = mySettings.Automatic_Household_Naming__c;
+        settings.Affiliation_Record_Type_Enforced__c = mySettings.Affiliation_Record_Type_Enforced__c;
```
[the remark this answers is not in the document]
Comment (xinnanlisf) [reply src/classes/UTIL_CustomSettingsFacade.cls]: @lparrott good catch!

**Comment by xinnanlisf** [comment]
@lparrott should be good again!

File: src/classes/AFFL_MultiRecordType_TDTM.cls [hunk 3]
```diff
@@ -101,7 +101,6 @@ public class AFFL_MultiRecordType_TDTM extends TDTM_Runnable {
 		                if (afflOld.Primary__c
 								&& !affl.Primary__c
 								&& lookupFieldName != null
-								&& UTIL_Describe.isValidField('Contact', lookupFieldName)
```
Comment (lparrott) [inline src/classes/AFFL_MultiRecordType_TDTM.cls]: @xinnanlisf I'm curious why this is failing for namespace builds, but not managed package builds. Are we passing a fieldname as a string, so it's not getting the namespace added correctly?

Comment (xinnanlisf) [reply src/classes/AFFL_MultiRecordType_TDTM.cls]: @lparrott great question! It looks like Util_Describe.isValidField is failing:

`    public static boolean isValidField(String objectName, String fieldName) {
        // fields in our own package must not have their prefix for the Describe Field Map
       fieldName = UTIL_Namespace.StrTokenRemoveNSPrefix(fieldName);

        // make sure we have this field's schema mapped
            if (!fieldDescribes.containsKey(objectName) || !fieldDescribes.get(objectName).containsKey(fieldName))
                fillFieldMapsForObject(objectName, fieldName);

            Schema.DescribeFieldResult dfr = fieldDescribes.get(objectName).get(fieldName);
            return dfr != null;
        
    }`

If we remove fieldName = UTIL_Namespace.StrTokenRemoveNSPrefix(fieldName);, that unit test would pass. It sounds like namespace scratch org does not think primary__household__c is a valid field. Rather, it thinks hed__primary__household__c is valid.

Comment (lparrott) [reply src/classes/AFFL_MultiRecordType_TDTM.cls]: Gotcha, thanks for looking into it! We can chalk that up to just a quirk of namespace scratch orgs (which should never exist anyways)

**Comment by xinnanlisf** [comment]
@lparrott i made some changes so we check if contact field is not null at the very beginning of the code. Could we take a review please? Thank you!

**Comment by lparrott** [comment]
@xinnanlisf The code looks fine as long as it works as expected in QA!

**Comment by annejackson** [comment]
I sent @xinnanlisf  some messages on this one with some screenshots. Thanks @juliesnow  for the changes/note help!

**Comment by xinnanlisf** [comment]
@lparrott this PR should be ready for review now. We are handling scenario when contact primary field in mappings mismatch.

**Comment by annejackson** [comment]
Found some issues on this, let @xinnanlisf know.

File: src/classes/AFFL_AccChange_TDTM.cls [hunk 4]
```diff
@@ -80,11 +80,13 @@ public with sharing class AFFL_AccChange_TDTM extends TDTM_Runnable {
...
                    }
 
-                   if (isAfflTypeEnforced) {
-                       ERR_ExceptionHandler.handleAfflException(affl, lookupFieldName);
-                   } else if (lookupFieldName == null || lookupFieldNameOld == null) {
-                       return dmlWrapper;
-                   }
+                    if (isAfflTypeEnforced) {
+                        ERR_ExceptionHandler.handleAfflNullRecordTypeException(affl);
+                    } else {
```
Comment (lparrott) [inline src/classes/AFFL_AccChange_TDTM.cls]: Why the change from `else if()`? I find the old way easier to read, but that's more of code styling question.

File: src/classes/AFFL_ContactChange_TDTM.cls [hunk 5]
```diff
@@ -82,11 +82,13 @@ public with sharing class AFFL_ContactChange_TDTM extends TDTM_Runnable {
...
 	               String lookupFieldName = AFFL_MultiRecordType_TDTM.afflMapper.contactLabelNames.get(lookupFieldLabel);
 
-				   if (isAfflTypeEnforced) {
-				       ERR_ExceptionHandler.handleAfflException(affl, lookupFieldName);
-				   }  else if (lookupFieldName == null) {
-					   return dmlWrapper;
-				   }
+					if (isAfflTypeEnforced) {
+						ERR_ExceptionHandler.handleAfflNullRecordTypeException(affl);
+					} else {
```
Comment (lparrott) [inline src/classes/AFFL_ContactChange_TDTM.cls]: Same as above.

File: src/classes/AFFL_MultiRecordType_TDTM.cls [hunk 6]
```diff
@@ -81,6 +81,12 @@ public class AFFL_MultiRecordType_TDTM extends TDTM_Runnable {
 						ERR_ExceptionHandler.handleAfflException(affl, lookupFieldName);
 					}  else if (lookupFieldName == null) {
 						return dmlWrapper;
+				 	if (isAfflTypeEnforced) {
+						ERR_ExceptionHandler.handleAfflNullRecordTypeException(affl);
+					} else {
```
Comment (lparrott) [inline src/classes/AFFL_MultiRecordType_TDTM.cls]: One more potential `else if()`

File: src/classes/AFFL_AccChange_TDTM.cls (hunk 4 above)
Comment (xinnanlisf) [reply src/classes/AFFL_AccChange_TDTM.cls]: Hey @lparrott, well IMO else if will get a little trickier if the conditional clause gets more complicated. Maybe we can have a standard for this thou.

Comment (lparrott) [reply src/classes/AFFL_AccChange_TDTM.cls]: Good point, it's easier for debugging too, let's leave your updates as is.

**Comment by xinnanlisf** [comment]
Hey @lparrott, I made some changes to error handling in last two commit. Mind taking a look?

**Comment by lparrott** [comment]
@xinnanlisf Great solution with using `if(SUBSTRINGS.keySet().contains(STATUS))` to get around that error you were seeing, this all looks fine to me.

**Comment by xinnanlisf** [comment]
Hey @ceiroa & @lparrott, I just made some polishment so that we don't hardcode anymore. (or less hardcoded) Let me know if that makes sense.

**Comment by annejackson** [comment]
This looks great @xinnanlisf!! Thanks for all the hard work. All my scenarios have passed. @ceiroa  and @lparrott let me know when one of you is done checking the polishments then I will merge.

**Comment by lparrott** [comment]
@xinnanlisf @annejackson Makes sense to me!

**Comment by annejackson** [comment]
Merging!

**Comment by xinnanlisf** [comment]
@juliesnow I moved the second paragraph and added a sub bullet. Let me know if it makes sense. Thank you!

**Comment by annejackson** [comment]
I like the sub-bullet, it looks good with all the release notes together.

**Comment by xinnanlisf** [comment]
Made some minor changes to release note based on @juliesnow's feedback.

**Comment by annejackson** [comment]
Updated release notes based on findings.
